[Feature] 코디 등록 UI 구현

dudtjs1021ej opened the pull request:
### 📕 Issue

x

### 📙 작업 내역

- [x] 코디 등록 UI
- [x] 앨범 UI
- [x] 전체 사진 불러오기 구현(앨범 목록 디자인 나오면 추가 구현 예정)
- [x] Common CollectionView Header 구현
- [x] Notification.Name + extension 추가
- [x] Instagram버튼, 앨범선택버튼, x버튼 이미지 추가


### 📘 작업 유형

- [x] 신규 기능 추가
- [ ] 버그 수정
- [ ] 리펙토링
- [ ] 문서 업데이트


### 📋 체크리스트

- [x] Merge 하는 브랜치가 올바른가?
- [x] PR과 관련없는 변경사항이 없는가?
- [x] 내 코드에 대한 자기 검토가 되었는가?
- [ ] 변경사항이 효과적이거나 동작이 작동한다는 것을 보증하는 테스트를 추가하였는가?
- [ ] 새로운 테스트와 기존의 테스트가 변경사항에 대해 만족하는가?

### 📱 테스트 방법

x

### 📝 리뷰 노트

- 코디네이터 패턴 맞게 적용했는지 한 번 봐주세요..! (특히 UploadCodyCoordinator에서 dismiss하는 부분😂)
- `CommonHeaderView` 구현해보았습니다. (구현하면서 SwiftLint 파라미터 개수 disable했습니다.)
- 코디등록 UI와 앨범 UI 레이아웃 `CollectionViewCompositionalLayout`으로 구현헸습니다. 첨 써봐서 미흡한 부분 바로 말해주세요!
- 사진 불러오는건 Photo 프레임워크를 사용해 `PhotoService`를 따로 구현했습니다. ([참고](https://ios-development.tistory.com/1024))
- (업로드 버튼은 공통 컴포넌트 나오면 수정하겠습니당)

### 🏞️ 스크린샷

<image src = "https://user-images.githubusercontent.com/77915491/212461410-58da6730-ed54-4c40-ab87-32e4846d95c6.png" width="40%"> <image src = "https://user-images.githubusercontent.com/77915491/212461462-f0ecd24b-49c8-403e-8b20-39ea673f8723.png" width="40%">

dudtjs1021ej committed:
✨ [feat] UploadCodyCoordinator 구현

dudtjs1021ej committed:
Merge branch 'develop' of https://github.com/YAPP-Github/21st-iOS-Team-2-iOS into feature/uploadCody-ios2-80

dudtjs1021ej committed:
✨ [feat] 코디 등록 상단 뷰 구현

- xButton 이미지 추가
- UploadCodyCoordinatorInterface dismissUploadCody 추가
- TopBarView 구현

dudtjs1021ej committed:
✨ [feat] UploadCodyViewController의 ContentCell 구현

- 인스타그램 버튼 이미지 추가
- CollectionView를 사용해 ContentCell 구현

dudtjs1021ej committed:
✨ [feat] StyleTagCell 구현

dudtjs1021ej committed:
✨ [feat] WeatherTagCell 구현

dudtjs1021ej committed:
♻️ [refactor] 코디 등록 상단 바 수정

- naivgationBar 추가
- TopBarView 삭제

dudtjs1021ej committed:
✨ [feat] Common HeaderView 구현

- Common HeaderView 구현
- swiftLint 파라미터 개수 비활성화

dudtjs1021ej committed:
✨ [feat] 스타일태그, 날씨태그 선택된 셀 컬러 삽입

- enum StyleTag, WeatherTag 추가

dudtjs1021ej committed:
🚚 [style] MainFeed 구조 변경

dudtjs1021ej committed:
✨ [feat] AlbumCoordinator 추가

dudtjs1021ej committed:
✨ [feat] 앨범 BarView 구현

- xGray버튼 이미지 추가
- BarView 구현

dudtjs1021ej committed:
✨ [feat] 앨범 CollectionView 구현

- btnAlbumUnSelcted, btnAlbumUnSelcted 이미지 추가

dudtjs1021ej committed:
✨ [feat] 앨범 사진 가져오기 구현

dudtjs1021ej committed:
✨ [feat] textView 키보드 가림 현상 해결

- Notification.Name+extension 추가

dudtjs1021ej committed:
💄 [style] AutoLayout 수정

dudtjs1021ej committed:
🐛 [fix] TextView text 사라짐 해결

dudtjs1021ej committed:
♻️ [refactor] 스타일태그, 날씨태그 관련 코드 수정

- 태그 관리 메소드에서 프로퍼티로 변경
- collectionView color 추가
- 태그 레이아웃 수정

leeari95 committed:
🔧 [chore] ci_post_clone 쉘 스크립트가 sh가 아닌 bash에서 실행되도록 변경

leeari95 commented on Fitfty/Projects/Coordinator/Sources/Main/UploadCodyCoordinator.swift:
```diff
@@ -0,0 +1,67 @@
+        let viewController = makeUploadCodyViewController()
+        navigationController.navigationBar.prefersLargeTitles = true
+        navigationController.navigationBar.largeTitleTextAttributes =
+        [NSAttributedString.Key.font: FitftyFont.appleSDBold(size: 24).font ?? UIFont.systemFont(ofSize: 24)]
```
[P3] 이 부분은 코디네이터 내부보다는 ViewController 내부에서 설정해주는 것이 좋지 않을까용?

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/11/commits/2018ddf28d54dac0b22ae72dfc704d1a13b8f387
수정 완료👍

dudtjs1021ej committed:
🔧 [chore] swiftlint empty_count 규칙 제외

leeari95 commented on Fitfty/Projects/MainFeed/Sources/UploadCody/ViewControllers/AlbumViewController.swift:
```diff
@@ -0,0 +1,182 @@
+    
+    private let navigationBarView = BarView()
+    
+    private let bottomView = UIView()
```
[질문/P5] 바텀뷰가 존재해야하는 이유가 궁금합니다!
이 bottomView를 활용해서 레이아웃을 잡고 있는 것 같은데, 제거해보면 어떨까요?
레이아웃을 잡을 때, 버튼과 컬렉션뷰 만으로도 가능하니 좀더 개선해볼 수 있을 것 같습니다 😊

dudtjs1021ej replied:
CollectionView와 구분해주기위해 따로 바텀뷰를 넣어주었습니다!
불필요한 UIView가 올라와있는 느낌일 수 있겠네요..! 수정하겠습니당

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/11/commits/2018ddf28d54dac0b22ae72dfc704d1a13b8f387
수정 완료👍

leeari95 commented on Fitfty/Projects/MainFeed/Sources/UploadCody/ViewControllers/AlbumViewController.swift:
```diff
@@ -0,0 +1,182 @@
+        view.addSubviews(navigationBarView, collectionView, bottomView, uploadButton)
+        
+        let collectionViewTopConstraint = collectionView.topAnchor.constraint(equalTo: navigationBarView.bottomAnchor)
+        collectionViewTopConstraint.priority = .defaultLow
```
[질문] 컬렉션뷰의 탑 우선순위를 낮춰주셨는데, 어떤 이유로 이렇게 설정해주신건지 궁금해요 😁

dudtjs1021ej replied:
navigationBarView의 heightAnchor와 충돌이 나서 해결하기 위해 넣었습니다!

leeari95 commented on Fitfty/Projects/MainFeed/Sources/UploadCody/ViewControllers/AlbumViewController.swift:
```diff
@@ -0,0 +1,182 @@
+        
+        NSLayoutConstraint.activate([
+            navigationBarView.topAnchor.constraint(equalTo: view.topAnchor),
+            navigationBarView.leftAnchor.constraint(equalTo: view.leftAnchor),
+            navigationBarView.rightAnchor.constraint(equalTo: view.rightAnchor),
```
[질문] leadingAnchor, trailingAnchor가 아니라 left, right를 사용해주신 이유가 궁금합니다!

dudtjs1021ej replied:
아 이 부분 실수했네여...😨 leading과 trailing으로 수정하겠습니다!

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/11/commits/556e203fc524b1c28dd402793ce3b1c86bc95caf
수정 완료😄

leeari95 commented on Fitfty/Projects/MainFeed/Sources/UploadCody/ViewControllers/AlbumViewController.swift:
```diff
@@ -0,0 +1,182 @@
+        case main
+    }
+    
+    private let navigationBarView = BarView()
```
[질문] 네비게이션 바를 사용할 수도 있을 것 같은데, BarView로 대체해주신 이유가 궁금하네요🤔

dudtjs1021ej replied:
디자인에 따르면 navigationBar 크기를 늘려야 하는데요
찾아보니 iOS11부터는 height값 변경이 되지 않는 것 같더라구요🥲 그래서 navigationBar를 hidden처리하고 따로 BarView를 만들었습니다.
혹시 좋은 방법 있으면 알려주세여..!!

leeari95 replied:
음... 네비게이션 바는 그대로 이용하고 콜렉션뷰 상단에 padding을 좀 주면 되지 않을까용!? (이상하려나...?)

dudtjs1021ej replied:
오옹..생각못한 방법이긴 한데 그럼 navigationItem이 가운데 정렬된 것 처럼 보여야해서 레이아웃 잡기가 좀 까다로울 것 같네요😂

leeari95 replied:
가운데 정렬이라면 수평 부분을 말씀하시는 걸까요?
ㅋㅋㅋ 저도 약간 비슷한 케이스로... 주소설정 화면 구현할때.. 
서치바 높이 조절이 공수가 많이들 것 같다고 레슬리한태 말씀드리니까...
어느정도 타협보긴 했거든요. 참고해주세용 ㅎㅎ (그래서 피그마랑 검색바 높이가 달라요 ㅋㅋ)
<img width="291" alt="image" src="https://user-images.githubusercontent.com/75905803/212812978-563d8082-875f-4e60-95b0-aee2c0c7ed2b.png">

dudtjs1021ej replied:
<image src = "https://user-images.githubusercontent.com/77915491/212847027-785db965-3928-405e-85fd-d9c795cea292.png" width="40%">
아리가 말해주신대로 구현 시도해봤는데요...!! 커스텀 바텀시트라 그런지 네비게이션바의 height가 너무 작아서 디자인대로는 나오기 힘들어보여요..ㅠㅠ (쉽게 보기 위해 navigationBar color를 brown으로 해두었어요!)
수직 중앙정렬도 구현이 좀 까다로워보이구요😭 이 부분은 레슬리와 얘기 나눠도 될 것 같구 일단은 그냥 두는 쪽으로 가도 될까요..!?🥹

leeari95 replied:
넵!! 어쩔수 없네요. 그게 좋을 것 같습니다! ㅠㅠ🥲
담에 한번 같이 개선해봐용!!!! ☺️💪

leeari95 commented on Fitfty/Projects/MainFeed/Sources/UploadCody/ViewControllers/AlbumViewController.swift:
```diff
@@ -0,0 +1,182 @@
+        return button
+    }()
+    
+    private lazy var collectionView: UICollectionView = {
+        let collectionView = UICollectionView(frame: .zero, collectionViewLayout: albumLayout())
+        collectionView.backgroundColor = .white
+        collectionView.showsVerticalScrollIndicator = false
+        collectionView.register(AlbumCell.self)
+        return collectionView
+    }()
```
[P3] 제가 직접 실행해보니까, 셀을 터치했을 때, 사진을 선택했다는 액션이 취해지지가 않던데,
이 부분도 UI를 구현할 때 같이 구현해두면 좋지 않을까요?
단순히 셀을 클릭했을 때 사진 체크만 표시되도록...? (UI 부분만요)
그리고 셀 내부에 버튼이 있던데, 셀의 버튼을 터치하지 않아도, 즉 셀만 터치해도 선택이 되면 좋을 것 같아요.

dudtjs1021ej replied:
이 부분은 피그마에서 선택 버튼 이미지가 export 되지 않아 구현하지 않았었는데 
이번에 아리가 주신 파일로 셀 터치로 구현해볼게요👍

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/11/commits/f8d2f721d75be5b669b4c9022ba501e282ab9e87
사진 선택 구현했습니다!

leeari95 commented on Fitfty/Projects/MainFeed/Sources/UploadCody/ViewControllers/AlbumViewController.swift:
```diff
@@ -0,0 +1,182 @@
+    
+    public override func viewDidLoad() {
+        super.viewDidLoad()
+        setNavigationBar()
+        setConstraintsLayout()
+        setDataSource()
+        applySnapshot()
+        setPhotoService()
```
[P3] 이 메서드들을 setUp() 이라는 메소드에 담아서 따로 분리해주고 viewDidLoad에서는 setUp() 만 호출할 수 있도록 해주면
더 깔끔해질 것 같습니다!

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/11/commits/2018ddf28d54dac0b22ae72dfc704d1a13b8f387
수정 완료👍

leeari95 commented on Fitfty/Projects/MainFeed/Sources/UploadCody/ViewControllers/AlbumViewController.swift:
```diff
@@ -0,0 +1,182 @@
+    
+    private var phAssets = [PHAsset]() {
+        didSet {
+            applySnapshot()
```
[P3] applySnapshot 메소드에 파라미터로 PHAsset을 넘겨주는 방식은 어떤가요?

dudtjs1021ej replied:
파라미터로 넘겨준다기보다는 phAssets이 값이 변경될때마다(ex 앨범 목록을 선택했을 때) applySnapshot을 호출하는 방식입니다! 이 부분도 수정이 필요할까요?

dudtjs1021ej replied:
혹시 didSet으로 호출하는 것보다 파라미터로 넘겨주는 방식을 선호하는 이유가 있을까요…!?

leeari95 replied:
나중에 뷰모델을 만들어서 넣어둘 데이터... 라고 생각이 들어서요!
파라미터로 넘겨주는 방식이라면, 추후 뷰모델을 구현한다고 했을 때
나중에 `phAssets`이라는 배열 데이터를 output으로 ViewController한태 넘겨주는 구조가 될 것 같아요.

어쨌든 구조상 '이 배열을 뷰모델이 가지고 있지 않을까?' 라는 생각이 들어서요.

지금은 setDataSource() 메소드 내부에서 이 배열을 활용해서... index로 접근하는 방식인 것 같은데,
이것도 수정해야하지 않을까요? 
데이터소스 타입이 `Section, UUID` 가 아니라 `Section, PHAsset`이 되야되지 않을까?! 했어요.
그럼 데이터소스로 셀을 설정할 때도 index 접근하는 방식도 필요없어질 것 같아요.

그래서 나중에 phAssets 이라는  output을 바인딩 해서 applySnapshot(phAssets)으로 넘겨주면,
appendItems를 호출할 때 임의적으로 
지금처럼 Array 이니셜라이저를 호출하는 게 아니라
바로 파라미터로 받은 배열을 넘겨주는 방향으로 가면 좋을 것 같아서 제안드려보았어요!

꼭 지금 수정하지 않아도 되요. 그냥 저의 생각은 이렇다라고 말씀드려요! 😊
나중에 추후 뷰모델을 구현할 때 다시 생각해봐도 좋아요 ㅎㅎ

dudtjs1021ej replied:
아아 그러네요..!!! viewModel 구현할때를 생각하면 파라미터로 넘기는 게 맞겠네요!
설명 감사해요 아리👍 수정해볼게요~!!

+요 부분은 다시 생각해보니 viewModel 구현할때 전체적으로 수정할 부분이 많아보여서 그때 한번에 수정하도록 하겠습니다!

leeari95 commented on Fitfty/Projects/Coordinator/Sources/Main/UploadCodyCoordinator.swift:
```diff
@@ -0,0 +1,67 @@
+        return viewController
+    }
+    
+    func makeAlbumViewController() -> UIViewController {
+        let coordinator = AlbumCoordinator()
+        coordinator.parentCoordinator = self
+        childCoordinators.append(coordinator)
+        coordinator.start()
+        let bottomSheetViewController = BottomSheetViewController(
+            style: .large,
+            contentViewController: coordinator.navigationController
+        )
+        return bottomSheetViewController
+        
+    }
```
[P5] 이 부분 제가 바텀시트를 조금 수정하는 바람에 추후 개선이 필요할 것 같습니다. 참고해주세요!
(메모리 누수 관련...)

leeari95 commented on Fitfty/Projects/MainFeed/Sources/UploadCody/Views/Album/BarView.swift:
```diff
@@ -0,0 +1,71 @@
+    private func setConstraintsLayout() {
+        addSubviews(albumStackView, cancelButton)
+        albumStackView.addArrangedSubviews(albumNameLabel, chevronButton)
+        albumStackView.spacing = 10
```
[P3] 16번 라인에 55번 라인과 같은 코드가 존재하네요!
제거해주셔도 될 것 같습니다 😊

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/11/commits/2018ddf28d54dac0b22ae72dfc704d1a13b8f387
수정 완료👍

leeari95 reviewed (changes requested):
안녕하세요. 썬 이번에도 UI 구현해주시느라 고생많으셨습니다!
궁금한 점 및 제 의견들을 코멘트로 남겨드렸는데, 혹시 더 리뷰가 필요한 곳이 있다면
꼭 말씀해주세요! 그리고 바텀시트에 관해서는 따로 음성으로 이야기를 나누면 더 좋을 것 같습니다!
시간 나실 때 이야기해주세요 😊🙏🏻
감사합니다!

leeari95 commented on Fitfty/Projects/Common/Resources/Images.xcassets/btnX.imageset/Contents.json:
```diff
@@ -0,0 +1,15 @@
+{
+  "images" : [
+    {
+      "filename" : "btnX.svg",
+      "idiom" : "universal"
+    }
+  ],
+  "info" : {
+    "author" : "xcode",
+    "version" : 1
+  },
+  "properties" : {
+    "preserves-vector-representation" : true
+  }
+}
```
[P5] 엇 이 버튼은 systemName으로도 불러올 수 있어요! 기본 아이콘이거든요!	
<img width="1077" alt="image" src="https://user-images.githubusercontent.com/75905803/212523484-1eebea81-64f4-4672-a680-3d6fd7017667.png">

dudtjs1021ej replied:
헙 이제 알았네여..!!🫢 수정하겠습니당

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/11/commits/7ce26ae750ef9cbaf4a104f85931d70328ddc98e
수정 완료👍

leeari95 commented on Fitfty/Projects/Common/Sources/HeaderView.swift:
```diff
@@ -0,0 +1,71 @@
+    private lazy var largeTitleLabel: UILabel = {
+        let label = UILabel()
+        label.textColor = .label
+        label.setContentHuggingPriority(.defaultHigh, for: .horizontal)
```
[P5] large, smal Label 둘다 HuggingPriority를 설정해주신 이유가 따로 있을까요?

dudtjs1021ej replied:
엇 지금 보니 불필요한 코드네요😅 지우겠습니다!!

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/commit/2018ddf28d54dac0b22ae72dfc704d1a13b8f387
수정 완료👍

leeari95 commented on Fitfty/Projects/MainFeed/Sources/UploadCody/Model/AlbumInfo.swift:
```diff
@@ -0,0 +1,16 @@
+struct AlbumInfo: Hashable {
+    let identifier: String?
+    let name: String
+    let count: Int
```
[P5] 이 count는 사진 갯수를 의미하는 건가요?
그렇다면 좀더 명확한 네이밍이 좋을 것 같아요!

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/commit/2018ddf28d54dac0b22ae72dfc704d1a13b8f387
`photoCount`로 수정했어요!

leeari95 commented on Fitfty/Projects/MainFeed/Sources/UploadCody/ViewControllers/AlbumViewController.swift:
```diff
@@ -0,0 +1,182 @@
+    
+    private let bottomView = UIView()
+    
+    private lazy var uploadButton: UIButton = {
+        let button = UIButton()
+        button.setTitle("업로드", for: .normal)
+        button.backgroundColor = .black
+        return button
+    }()
```
[P3] 업로드 버튼이 비활성화되는 부분도 구현해두면 좋을 것 같습니다!

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/11/commits/f8d2f721d75be5b669b4c9022ba501e282ab9e87
나중에 공통 컴포넌트 나오면 적용할 수 있게 setUploadButton 메소드 따로 만들었어여!

dudtjs1021ej commented on Fitfty/Projects/Coordinator/Sources/Main/UploadCodyCoordinator.swift:
```diff
@@ -0,0 +1,67 @@
+}
+
+extension UploadCodyCoordinator: UploadCodyCoordinatorInterface {
+    func dismissUploadCody(_ viewController: UIViewController) {
+        finishDelegate?.coordinatorDidFinish(childCoordinator: self)
+        viewController.dismiss(animated: true)
+    }
```
@leeari95  아리 많은 양 꼼꼼히 리뷰해주셔서 감사합니다! 리뷰해주신 부분 모두 반영해볼게요😆 추가적으로 이렇게 ViewController를 받아와서 dismiss 처리해줘도 되는지..? 궁금합니다!

leeari95 replied:
음 저는 별문제 없다고 생각합니당!
저는 보통 네비게이션 컨트롤러의 [visibleviewcontroller](https://developer.apple.com/documentation/uikit/uinavigationcontroller/1621862-visibleviewcontroller) 프로퍼티를 활용해서 dismiss 하기도 해요!
요놈이 [topViewController](https://developer.apple.com/documentation/uikit/uinavigationcontroller/1621849-topviewcontroller) 이랑은 쪼금 달라요. 
참고하시라고 공식문서 링크 연결했어요 ㅎㅎㅎ
그냥 이런것도 있구나~ 해주시면 됩니당 😁
썬이 구현한 방식은 파라미터로 viewController를 받아와서 더 확실한[?] 느낌인 것 같아요!

dudtjs1021ej replied:
오옹 이렇게 구현해도 상관없군요..!!  visibleViewController는 몰랐는데 링크까지..! 감사합니당ㅎㅎ 참고해볼게요😁

dudtjs1021ej committed:
♻️ [refactor] leftAnchor, rightAnchor를 leadingAnchor, trailingAnchor로 변경

dudtjs1021ej committed:
♻️ [refactor] 코드 개선 및 불필요한 코드 삭제

- navigationController 관련 코드 ViewController로 이동
- AlbumViewController BottomView 삭제
- AlbumViewController setUp 메소드 추가
- stackView spacing 중복 코드 제거
- HeaderView priority 코드 제거
- AlbumInfo 프로퍼티명 수정

dudtjs1021ej committed:
💄 [style] xBtn 이미지 삭제, systemImage 사용

dudtjs1021ej committed:
✨ [feat] AlbumCell 선택 구현

- btnAlbumSelected 이미지 추가

leeari95 committed:
Merge branch 'develop' into feature/uploadCody-ios2-80

leeari95 reviewed (approved):
수정사항 모두 확인했습니다.
제 의견 적극 반영해주셔서 너무 감사합니다!!! 썬 최고!!😊
고생많으셨습니다~~~☺️
